File: src/main/java/org/cobbzilla/util/io/ByteLimitedInputStream.java
package org.cobbzilla.util.io;

import java.io.IOException;
import java.io.InputStream;

public class ByteLimitedInputStream extends InputStream {

    private InputStream delegate;

    public long getCount() {
        return this.count;
    }

    public long getLimit() {
        return this.limit;
    }

    private interface BLISDelegateExcludes {
        int read(byte[] b) throws IOException;
        int read(byte[] b, int off, int len) throws IOException;
        int read() throws IOException;
    }

    private long count = 0;
    private long limit;

    public double getPercentDone () { return ((double) count) / ((double) limit); }

    public ByteLimitedInputStream (InputStream in, long limit) {
        this.delegate = in;
        this.limit = limit;
    }

    @Override public int read(byte[] b) throws IOException {
        if (count >= limit) return -1;
        final int read = delegate.read(b);
        if (read != -1) count += read;
        return read;
    }

    @Override public int read(byte[] b, int off, int len) throws IOException {
        if (count >= limit) return -1;
        final int read = delegate.read(b, off, len);
        if (read != -1) count += read;
        return read;
    }

    @Override public int read() throws IOException {
        if (count >= limit) return -1;
        final int read = delegate.read();
        if (read != -1) count++;
        return read;
    }

}

File: src/main/java/org/cobbzilla/util/io/ByteLimitedInputStream.kt
package org.cobbzilla.util.io

import java.io.IOException
import java.io.InputStream

class ByteLimitedInputStream(private val delegate: InputStream, val limit: Long) : InputStream() {

    var count: Long = 0
        private set

    val percentDone: Double
        get() = count.toDouble() / limit.toDouble()

    private interface BLISDelegateExcludes {
        @Throws(IOException::class)
        fun read(b: ByteArray): Int

        @Throws(IOException::class)
        fun read(b: ByteArray, off: Int, len: Int): Int

        @Throws(IOException::class)
        fun read(): Int
    }

    @Throws(IOException::class)
    override fun read(b: ByteArray): Int {
        if (count >= limit) return -1
        val read = delegate.read(b)
        if (read != -1) count += read.toLong()
        return read
    }

    @Throws(IOException::class)
    override fun read(b: ByteArray, off: Int, len: Int): Int {
        if (count >= limit) return -1
        val read = delegate.read(b, off, len)
        if (read != -1) count += read.toLong()
        return read
    }

    @Throws(IOException::class)
    override fun read(): Int {
        if (count >= limit) return -1
        val read = delegate.read()
        if (read != -1) count++
        return read
    }

}
